packages/flutter_tools/lib/src/ios/core_devices.dart

```diff
@@ -0,0 +1,847 @@
+  /// run the command.
+  static const int _minimumTimeoutInSeconds = 5;
+
+  /// Executes `devicectl` command to get list of devices. The command will
+  /// likely complete before [timeout] is reached. If [timeout] is reached,
+  /// the command will be stopped as a failure.
+  Future<List<Object?>> _listCoreDevices({
+    Duration timeout = const Duration(seconds: _minimumTimeoutInSeconds),
+  }) async {
+    if (!_xcode.isDevicectlInstalled) {
+      _logger.printError('devicectl is not installed.');
+      return <Object?>[];
+    }
+
+    // Default to minimum timeout if needed to prevent error.
+    Duration validTimeout = timeout;
+    if (timeout.inSeconds < _minimumTimeoutInSeconds) {
+      _logger.printError(
+          'Timeout of ${timeout.inSeconds} seconds is below the minimum timeout value '
+          'for devicectl. Changing the timeout to the minimum value of $_minimumTimeoutInSeconds.');
+      validTimeout = const Duration(seconds: _minimumTimeoutInSeconds);
+    }
+
+    final Directory tempDirectory = _fileSystem.systemTempDirectory.createTempSync('core_devices.');
+    final File output = tempDirectory.childFile('core_device_list.json');
+    output.createSync();
+
+    final List<String> command = <String>[
+      ..._xcode.xcrunCommand(),
+      'devicectl',
+      'list',
+      'devices',
+      '--timeout',
+      validTimeout.inSeconds.toString(),
+      '--json-output',
+      output.path,
+    ];
+
+    try {
+      await _processUtils.run(command, throwOnError: true);
+
+      final String stringOutput = output.readAsStringSync();
+
+      try {
+        final Object? decodeResult = (json.decode(stringOutput) as Map<String, Object?>)['result'];
+        if (decodeResult is Map<String, Object?>) {
+          final Object? decodeDevices = decodeResult['devices'];
+          if (decodeDevices is List<Object?>) {
+            return decodeDevices;
+          }
+        }
+        _logger.printError('devicectl returned unexpected JSON response: $stringOutput');
+        return <Object?>[];
+      } on FormatException {
+        // We failed to parse the devicectl output, or it returned junk.
+        _logger.printError('devicectl returned non-JSON response: $stringOutput');
+        return <Object?>[];
+      }
+    } on ProcessException catch (err) {
+      _logger.printError('Error executing devicectl: $err');
+      return <Object?>[];
+    } finally {
+      tempDirectory.deleteSync(recursive: true);
+    }
+  }
+
+  Future<List<IOSCoreDevice>> getCoreDevices({
+    Duration timeout = const Duration(seconds: _minimumTimeoutInSeconds),
+  }) async {
+    final List<IOSCoreDevice> devices = <IOSCoreDevice>[];
+
+    final List<Object?> devicesSection = await _listCoreDevices(timeout: timeout);
+    for (final Object? deviceObject in devicesSection) {
+      if (deviceObject is Map<String, Object?>) {
+        devices.add(IOSCoreDevice.fromPreviewJson(deviceObject, logger: _logger));
```

Comment:
what is "preview" json

Reply:
Currently, `devicectl` is in "preview mode", it's not stable. If you do `xcrun devicectl --help`, it says:

>This tool is being provided in a preview format to give all users an opportunity to test out the tool and
provide feedback. Given the preview nature of the tool, we cannot guarantee that the command set and output
(including JSON output) will remain stable before the final release of the tool. A future version of devicectl
will provide a supported and stable set of subcommands and a stable API via JSON output captured to a file.

Reply:
Gotcha. How about "PreRelease" or "Beta"? Preview means too many things.